src/Rules/Arrays/ArrayDestructuringRule.php

```diff
@@ -72,7 +73,7 @@ static function (Type $varType): bool {
 		if ($exprType instanceof ErrorType) {
 			return [];
 		}
-		if (!$exprType->isArray()->yes()) {
+		if (!$exprType->isArray()->yes() && !(new ObjectType(\ArrayAccess::class))->isSuperTypeOf($exprType)->yes()) {
```

Comment:
Same thing needs to be repeated inside the `static function` a few lines above in the `findTypetoCheck` call.

Reply:
Thx! I was still trying to figure out if and how to test that change. I was currently tinkering around with a new test class where the checkUnionTypes of the ruleLevelHelper is disabled to verify that this works. Feel free to let me know if this is needed

Reply:
You can test this stuff with integration tests here: https://github.com/phpstan/phpstan-src/tree/master/tests/PHPStan/Levels

There are some test files and some JSONs. If you modify the test file and run a test where it's included (most are here): https://github.com/phpstan/phpstan-src/blob/master/tests/PHPStan/Levels/LevelsIntegrationTest.php, PHPStan will modify the JSON files to reflect the current state. The key is to realize which situations are reported on which level. In this case there are going to be different levels reported for trying to destructure `int`, `array|null`, `array|Foo`, `ArrayObject|null`, `ArrayObject|Foo` and that's where this change is important.